### PR TITLE
cmd/testwrapper: exit code 1 when go build fails

### LINKED ISSUE #8586: cmd/testwrapper: add tests for JSON decoding
Follow up from #8520

### LINKED ISSUE #9275: cmd/testwrapper: exits with code 0 despite build errors
### What is the issue?

`testwrapper ./builderror` exits with status code 0 even if the builderror package fails to compile.

### Steps to reproduce

```
$ mkdir builderror

$ echo $'package builderror\n\nderp\n' > builderror/builderror.go

$ go test ./builderror && echo test-ok || echo test-fail
# tailscale.com/builderror
builderror/builderror.go:3:1: syntax error: non-declaration statement outside function body
FAIL
test-fail

$ /tmp/testwrapper ./builderror && echo test-ok || echo test-fail
# tailscale.com/builderror
builderror/builderror.go:3:1: syntax error: non-declaration statement outside function body
test-ok
```

### Are there any recent changes that introduced the issue?

Appears to have been introduced in 8e840489ed0409abdcaa3b1942f379a9e6e05625:

```
$ go run tailscale.com/cmd/testwrapper@8e840489ed0409abdcaa3b1942f379a9e6e05625 ./builderror && echo test-ok || echo test-fail
# tailscale.com/builderror
builderror/builderror.go:3:1: syntax error: non-declaration statement outside function body
PASS
test-ok
```

The parent commit 2cf6e127907641bdb9eb5cd8e8cf14e968b571d7 would swallow the build failure output but exit with the correct code:

```
$ go run tailscale.com/cmd/testwrapper@2cf6e127907641bdb9eb5cd8e8cf14e968b571d7 ./builderror && echo test-ok || echo test-fail
exit status 1
test-fail
```

### OS

macOS

### OS version

_No response_

### Tailscale version

_No response_

### Other software

_No response_

### Bug report

_No response_

### PR DESCRIPTION
**Pull request by icio**
Note that merging this PR will then block upgrading tailscale.com in corp until a darwin build error from the PR below is fixed.

Fixes #9275
Fixes #8586
Fixes tailscale/corp#13115

### CODE REVIEW

File: cmd/testwrapper/testwrapper.go [hunk 1]
```diff
@@ -273,6 +278,17 @@ func main() {
 				}
 			}
 		}
+		if err := <-runErr; err != nil {
+			var exit *exec.ExitError
+			if errors.As(err, &exit) {
+				if code := exit.ExitCode(); code > -1 {
+					os.Exit(exit.ExitCode())
```
Comment (icio) [inline cmd/testwrapper/testwrapper.go:285]: This is in the wrong spot. We'll always get a nonzero exit code when there are flakey tests we want to retry.

Comment (icio) [reply cmd/testwrapper/testwrapper.go:285]: I've got this fixed but I think I'm going to write a bunch more test cases.

Comment (icio) [reply cmd/testwrapper/testwrapper.go]: Updated in https://github.com/tailscale/tailscale/pull/9276/commits/e1770713d58e44ad3947471d6071bbae78bcfc31

File: cmd/testwrapper/testwrapper_test.go [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+	}
+	return exec.Command(buildPath, args...)
+}
+
+func buildTestWrapper() (string, error) {
+	dir, err := os.MkdirTemp("", "testwrapper")
+	if err != nil {
+		return "", fmt.Errorf("making temp dir: %w", err)
+	}
+	_, err = exec.Command("go", "build", "-o", dir, ".").Output()
```
Comment (icio) [inline cmd/testwrapper/testwrapper_test.go:38]: Maybe this is a good cue to refactor `main` to be callable from the test.

File: cmd/testwrapper/testwrapper.go [hunk 3]
```diff
@@ -244,12 +249,21 @@ func main() {
...
-			go runTests(ctx, thisRun.attempt, pt, otherArgs, ch)
+			runErr := make(chan error, 1)
+			go func() {
+				defer close(runErr)
+				runErr <- runTests(ctx, thisRun.attempt, pt, otherArgs, ch)
+			}()
+
+			var failed bool
 			for tr := range ch {
+				if tr.pkg == "command-line-arguments" {
```
Comment (maisem) [inline cmd/testwrapper/testwrapper.go:263]: this should be documented somewhere higher up

Comment (icio) [reply cmd/testwrapper/testwrapper.go:267]: Documented in https://github.com/tailscale/tailscale/pull/9276/commits/b9d92b4ed64c1caa71883d9a264c9e5d17391276

**Review by maisem — APPROVED** [review]

File: cmd/testwrapper/testwrapper_test.go [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+	return filepath.Join(dir, "testwrapper"), nil
+}
+
+func TestMain(m *testing.M) {
+	code := m.Run()
+	if buildPath != "" {
+		os.Remove(buildPath)
+	}
+	os.Exit(code)
+}
```
Comment (maisem) [inline cmd/testwrapper/testwrapper_test.go:51]: i don't think you need this

**Pull request merged**